BlockingOperationNotAllowedException When using quarkus-test-security + quarkus.security.jaxrs.deny-unannotated-endpoints = true

### Describe the bug

It seems that `quarkus.security.jaxrs.deny-unannotated-endpoints = true` has some problems with `quarkus-test-security`.
This only seems to be a problem on Resources with exception mappers inside the class.


Stacktrace:

```
Suppressed: io.quarkus.runtime.BlockingOperationNotAllowedException: Blocking security check attempted in code running on the event loop. Make the secured method return an async type, i.e. Uni, Multi or CompletionStage, or use an authentication mechanism that sets the SecurityIdentity in a blocking manner prior to delegating the call
	at io.quarkus.security.runtime.interceptor.SecurityConstrainer.check(SecurityConstrainer.java:34)
	at io.quarkus.security.runtime.interceptor.SecurityHandler.handle(SecurityHandler.java:46)
	at io.quarkus.security.runtime.interceptor.DenyAllInterceptor.intercept(DenyAllInterceptor.java:30)
	at io.quarkus.security.runtime.interceptor.DenyAllInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:33)
	at com.dfrickert.rest.resources.func.RestService_Subclass.authenticationFailedHandler(Unknown Source)
	at com.dfrickert.rest.resources.func.RestService$GeneratedExceptionHandlerFor$UnauthorizedException$OfMethod$authenticationFailedHandler.toResponse(Unknown Source)
	at com.dfrickert.rest.resources.func.RestService$GeneratedExceptionHandlerFor$UnauthorizedException$OfMethod$authenticationFailedHandler.toResponse(Unknown Source)
	at org.jboss.resteasy.reactive.server.core.RuntimeExceptionMapper.mapException(RuntimeExceptionMapper.java:93)
	at org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext.mapExceptionIfPresent(ResteasyReactiveRequestContext.java:342)
	at org.jboss.resteasy.reactive.server.handlers.ExceptionHandler.handle(ExceptionHandler.java:15)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:124)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:142)
	at org.jboss.resteasy.reactive.server.handlers.RestInitialHandler.beginProcessing(RestInitialHandler.java:51)
	at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:18)
	at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:8)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:430)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:408)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.security.HttpAuthorizer.doPermissionCheck(HttpAuthorizer.java:117)
	at io.quarkus.vertx.http.runtime.security.HttpAuthorizer.checkPermission(HttpAuthorizer.java:100)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$3.handle(HttpSecurityRecorder.java:233)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$3.handle(HttpSecurityRecorder.java:225)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:206)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:60)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:68)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:37)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$15.handle(VertxHttpRecorder.java:606)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$15.handle(VertxHttpRecorder.java:589)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:185)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:160)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:67)
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:30)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:55)
	at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:158)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:145)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:157)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:116)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:336)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:444)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:280)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61)
	at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
	Caused by: io.quarkus.runtime.BlockingOperationNotAllowedException: Cannot call getIdentity() from the IO thread when lazy authentication is in use, as resolving the identity may block the thread. Instead you should inject the CurrentIdentityAssociation, call CurrentIdentityAssociation#getDeferredIdentity() and subscribe to the Uni.
	at io.quarkus.security.runtime.SecurityIdentityAssociation.getIdentity(SecurityIdentityAssociation.java:68)
	at io.quarkus.test.security.DelegateSecurityIdentityAssociation_ClientProxy.getIdentity(Unknown Source)
	at io.quarkus.test.security.TestIdentityAssociation.getIdentity(TestIdentityAssociation.java:71)
	at io.quarkus.test.security.TestIdentityAssociation_ClientProxy.getIdentity(Unknown Source)
	at io.quarkus.security.runtime.interceptor.SecurityConstrainer.check(SecurityConstrainer.java:32)
		... 100 more

```

### Expected behavior

Test returns appropriate status code depending on the test case (401/403)

### Actual behavior

Test fails with HTTP 500

### How to Reproduce?

https://github.com/davidfrickert/quarkus-issue-reproducer/tree/security-test-resource-exception-mappers-auth

Run test `GreetingResourceImplTest`. Expected 403, actually is 500 with the error on the issue description.

### Output of `uname -a` or `ver`

Linux dfrickert 5.15.74-4-MANJARO #1 SMP PREEMPT Sat Oct 15 18:49:48 UTC 2022 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.5" 2022-10-18 OpenJDK Runtime Environment (build 17.0.5+1) OpenJDK 64-Bit Server VM (build 17.0.5+1, mixed mode)

### GraalVM version (if different from Java)

N/A

### Quarkus version or git rev

2.13.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63) Maven home: /opt/maven Java version: 17.0.5, vendor: N/A, runtime: /usr/lib/jvm/java-17-openjdk Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.15.74-4-manjaro", arch: "amd64", family: "unix"

### Additional information

_No response_

--- a/extensions/resteasy-classic/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/RestPathAnnotationProcessor.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/RestPathAnnotationProcessor.java
@@ -1,6 +1,5 @@
 package io.quarkus.resteasy.deployment;
 
-import java.util.List;
 import java.util.Optional;
 import java.util.regex.Pattern;
 
@@ -81,18 +80,15 @@ public class RestPathAnnotationProcessor {
                 MethodInfo methodInfo = target.asMethod();
                 ClassInfo classInfo = methodInfo.declaringClass();
 
-                AnnotationInstance annotation = methodInfo.annotation(REST_PATH);
-                if (annotation == null) {
-                    // Check for @Path on class and not method
-                    if (!isRestEndpointMethod(methodInfo.annotations())) {
-                        return;
-                    }
+                if (!isRestEndpointMethod(methodInfo)) {
+                    return;
                 }
                 // Don't create annotations for rest clients
                 if (classInfo.classAnnotation(REGISTER_REST_CLIENT) != null) {
                     return;
                 }
 
+                AnnotationInstance annotation = methodInfo.annotation(REST_PATH);
                 StringBuilder stringBuilder;
                 if (annotation != null) {
                     stringBuilder = new StringBuilder(slashify(annotation.value().asString()));
@@ -138,17 +134,18 @@ public class RestPathAnnotationProcessor {
         return '/' + path;
     }
 
-    boolean isRestEndpointMethod(List<AnnotationInstance> annotations) {
-        boolean isRestEndpointMethod = false;
+    static boolean isRestEndpointMethod(MethodInfo methodInfo) {
 
-        for (AnnotationInstance annotation : annotations) {
-            if (ResteasyDotNames.JAXRS_METHOD_ANNOTATIONS.contains(annotation.name())) {
-                isRestEndpointMethod = true;
-                break;
+        if (!methodInfo.hasAnnotation(REST_PATH)) {
+            // Check for @Path on class and not method
+            for (AnnotationInstance annotation : methodInfo.annotations()) {
+                if (ResteasyDotNames.JAXRS_METHOD_ANNOTATIONS.contains(annotation.name())) {
+                    return true;
+                }
             }
+            return false;
         }
-
-        return isRestEndpointMethod;
+        return true;
     }
 
     private boolean notRequired(Capabilities capabilities,

--- a/extensions/resteasy-classic/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyBuiltinsProcessor.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyBuiltinsProcessor.java
@@ -1,6 +1,7 @@
 package io.quarkus.resteasy.deployment;
 
 import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
+import static io.quarkus.resteasy.deployment.RestPathAnnotationProcessor.isRestEndpointMethod;
 import static io.quarkus.security.spi.SecurityTransformerUtils.hasSecurityAnnotation;
 
 import java.util.ArrayList;
@@ -10,6 +11,7 @@ import java.util.stream.Collectors;
 
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
+import org.jboss.jandex.MethodInfo;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.deployment.Capabilities;
@@ -36,7 +38,7 @@ import io.quarkus.resteasy.runtime.vertx.JsonArrayWriter;
 import io.quarkus.resteasy.runtime.vertx.JsonObjectReader;
 import io.quarkus.resteasy.runtime.vertx.JsonObjectWriter;
 import io.quarkus.resteasy.server.common.deployment.ResteasyDeploymentBuildItem;
-import io.quarkus.security.spi.AdditionalSecuredClassesBuildItem;
+import io.quarkus.security.spi.AdditionalSecuredMethodsBuildItem;
 import io.quarkus.vertx.http.deployment.HttpRootPathBuildItem;
 import io.quarkus.vertx.http.deployment.devmode.NotFoundPageDisplayableEndpointBuildItem;
 import io.quarkus.vertx.http.deployment.devmode.RouteDescriptionBuildItem;
@@ -51,30 +53,31 @@ public class ResteasyBuiltinsProcessor {
     void setUpDenyAllJaxRs(CombinedIndexBuildItem index,
             JaxRsSecurityConfig config,
             ResteasyDeploymentBuildItem resteasyDeployment,
-            BuildProducer<AdditionalSecuredClassesBuildItem> additionalSecuredClasses) {
-        if ((config.denyJaxRs) && (resteasyDeployment != null)) {
-            final List<ClassInfo> classes = new ArrayList<>();
+            BuildProducer<AdditionalSecuredMethodsBuildItem> additionalSecuredClasses) {
+        if (resteasyDeployment != null && (config.denyJaxRs || config.defaultRolesAllowed.isPresent())) {
+            final List<MethodInfo> methods = new ArrayList<>();
 
+            // add endpoints
             List<String> resourceClasses = resteasyDeployment.getDeployment().getScannedResourceClasses();
             for (String className : resourceClasses) {
                 ClassInfo classInfo = index.getIndex().getClassByName(DotName.createSimple(className));
                 if (!hasSecurityAnnotation(classInfo)) {
-                    classes.add(classInfo);
+                    for (MethodInfo methodInfo : classInfo.methods()) {
+                        if (isRestEndpointMethod(methodInfo) && !hasSecurityAnnotation(methodInfo)) {
+                            methods.add(methodInfo);
+                        }
+                    }
                 }
             }
 
-            additionalSecuredClasses.produce(new AdditionalSecuredClassesBuildItem(classes));
-        } else if (config.defaultRolesAllowed.isPresent() && resteasyDeployment != null) {
-            final List<ClassInfo> classes = new ArrayList<>();
-
-            List<String> resourceClasses = resteasyDeployment.getDeployment().getScannedResourceClasses();
-            for (String className : resourceClasses) {
-                ClassInfo classInfo = index.getIndex().getClassByName(DotName.createSimple(className));
-                if (!hasSecurityAnnotation(classInfo)) {
-                    classes.add(classInfo);
+            if (!methods.isEmpty()) {
+                if (config.denyJaxRs) {
+                    additionalSecuredClasses.produce(new AdditionalSecuredMethodsBuildItem(methods));
+                } else {
+                    additionalSecuredClasses
+                            .produce(new AdditionalSecuredMethodsBuildItem(methods, config.defaultRolesAllowed));
                 }
             }
-            additionalSecuredClasses.produce(new AdditionalSecuredClassesBuildItem(classes, config.defaultRolesAllowed));
         }
     }
 

--- a/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/DefaultRolesAllowedJaxRsTest.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/DefaultRolesAllowedJaxRsTest.java
@@ -3,6 +3,11 @@ package io.quarkus.resteasy.test.security;
 import static io.restassured.RestAssured.given;
 import static io.restassured.RestAssured.when;
 
+import javax.annotation.security.RolesAllowed;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+import org.hamcrest.Matchers;
 import org.jboss.shrinkwrap.api.asset.StringAsset;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -19,7 +24,7 @@ public class DefaultRolesAllowedJaxRsTest {
                     .addClasses(PermitAllResource.class, UnsecuredResource.class,
                             TestIdentityProvider.class,
                             TestIdentityController.class,
-                            UnsecuredSubResource.class)
+                            UnsecuredSubResource.class, HelloResource.class)
                     .addAsResource(new StringAsset("quarkus.security.jaxrs.default-roles-allowed = admin\n"),
                             "application.properties"));
 
@@ -65,6 +70,17 @@ public class DefaultRolesAllowedJaxRsTest {
         assertStatus(path, 200, 200, 200);
     }
 
+    @Test
+    public void testNonEndpointMethodAreNotDenied() {
+        // ensure io.quarkus.resteasy.test.security.DefaultRolesAllowedJaxRsTest.HelloResource.getHello is not secured with RolesAllowedInterceptor
+        given().auth().preemptive()
+                .basic("user", "user")
+                .get("/hello")
+                .then()
+                .statusCode(200)
+                .body(Matchers.equalTo("hello"));
+    }
+
     private void assertStatus(String path, int adminStatus, int userStatus, int anonStatus) {
         given().auth().preemptive()
                 .basic("admin", "admin").get(path)
@@ -77,6 +93,21 @@ public class DefaultRolesAllowedJaxRsTest {
         when().get(path)
                 .then()
                 .statusCode(anonStatus);
+
+    }
+
+    @Path("/hello")
+    public static class HelloResource {
+
+        @RolesAllowed("**")
+        @GET
+        public String hello() {
+            return getHello();
+        }
+
+        public String getHello() {
+            return "hello";
+        }
 
     }
 

--- a/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/DenyAllJaxRsTest.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/security/DenyAllJaxRsTest.java
@@ -4,6 +4,11 @@ import static io.restassured.RestAssured.given;
 import static io.restassured.RestAssured.when;
 import static org.hamcrest.Matchers.emptyString;
 
+import javax.annotation.security.PermitAll;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+import org.hamcrest.Matchers;
 import org.jboss.shrinkwrap.api.asset.StringAsset;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -23,7 +28,7 @@ public class DenyAllJaxRsTest {
                     .addClasses(PermitAllResource.class, UnsecuredResource.class,
                             TestIdentityProvider.class,
                             TestIdentityController.class,
-                            UnsecuredSubResource.class)
+                            UnsecuredSubResource.class, HelloResource.class)
                     .addAsResource(new StringAsset("quarkus.security.jaxrs.deny-unannotated-endpoints = true\n"),
                             "application.properties"));
 
@@ -82,6 +87,16 @@ public class DenyAllJaxRsTest {
         assertStatus(path, 200, 200);
     }
 
+    @Test
+    public void testNonEndpointMethodAreNotDenied() {
+        // ensure io.quarkus.resteasy.test.security.DenyAllJaxRsTest.HelloResource.getHello is not secured with DenyAllInterceptor
+        given()
+                .get("/hello")
+                .then()
+                .statusCode(200)
+                .body(Matchers.equalTo("hello"));
+    }
+
     private void assertStatus(String path, int status, int anonStatus) {
         given().auth().preemptive()
                 .basic("admin", "admin").get(path)
@@ -94,6 +109,21 @@ public class DenyAllJaxRsTest {
         when().get(path)
                 .then()
                 .statusCode(anonStatus);
+
+    }
+
+    @Path("/hello")
+    public static class HelloResource {
+
+        @PermitAll
+        @GET
+        public String hello() {
+            return getHello();
+        }
+
+        public String getHello() {
+            return "hello";
+        }
 
     }
 

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java
@@ -1,10 +1,13 @@
 package io.quarkus.resteasy.reactive.common.deployment;
 
+import static io.quarkus.security.spi.SecurityTransformerUtils.hasSecurityAnnotation;
 import static org.jboss.resteasy.reactive.common.model.ResourceInterceptor.FILTER_SOURCE_METHOD_METADATA_KEY;
+import static org.jboss.resteasy.reactive.common.processor.EndpointIndexer.collectClassEndpoints;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
@@ -18,7 +21,6 @@ import javax.ws.rs.Priorities;
 import javax.ws.rs.ext.RuntimeDelegate;
 
 import org.jboss.jandex.AnnotationTarget;
-import org.jboss.jandex.AnnotationTarget.Kind;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.CompositeIndex;
 import org.jboss.jandex.DotName;
@@ -61,8 +63,7 @@ import io.quarkus.resteasy.reactive.spi.MessageBodyWriterBuildItem;
 import io.quarkus.resteasy.reactive.spi.MessageBodyWriterOverrideBuildItem;
 import io.quarkus.resteasy.reactive.spi.ReaderInterceptorBuildItem;
 import io.quarkus.resteasy.reactive.spi.WriterInterceptorBuildItem;
-import io.quarkus.security.spi.AdditionalSecuredClassesBuildItem;
-import io.quarkus.security.spi.SecurityTransformerUtils;
+import io.quarkus.security.spi.AdditionalSecuredMethodsBuildItem;
 
 public class ResteasyReactiveCommonProcessor {
 
@@ -72,33 +73,42 @@ public class ResteasyReactiveCommonProcessor {
     @BuildStep
     void setUpDenyAllJaxRs(
             CombinedIndexBuildItem index,
-            ResteasyReactiveConfig rrConfig,
             JaxRsSecurityConfig securityConfig,
             Optional<ResourceScanningResultBuildItem> resteasyDeployment,
-            BuildProducer<AdditionalSecuredClassesBuildItem> additionalSecuredClasses) {
+            BeanArchiveIndexBuildItem beanArchiveIndexBuildItem,
+            ApplicationResultBuildItem applicationResultBuildItem,
+            BuildProducer<AdditionalSecuredMethodsBuildItem> additionalSecuredClasses) {
 
-        if (securityConfig.denyJaxRs() && resteasyDeployment.isPresent()) {
-            List<ClassInfo> classes = new ArrayList<>();
+        if (resteasyDeployment.isPresent()
+                && (securityConfig.denyJaxRs() || securityConfig.defaultRolesAllowed().isPresent())) {
+            final List<MethodInfo> methods = new ArrayList<>();
+            Map<DotName, String> httpAnnotationToMethod = resteasyDeployment.get().getResult().getHttpAnnotationToMethod();
             Set<DotName> resourceClasses = resteasyDeployment.get().getResult().getScannedResourcePaths().keySet();
+
             for (DotName className : resourceClasses) {
                 ClassInfo classInfo = index.getIndex().getClassByName(className);
-                if (!SecurityTransformerUtils.hasSecurityAnnotation(classInfo)) {
-                    classes.add(classInfo);
+                if (!hasSecurityAnnotation(classInfo)) {
+                    // collect class endpoints
+                    Collection<MethodInfo> classEndpoints = collectClassEndpoints(classInfo, httpAnnotationToMethod,
+                            beanArchiveIndexBuildItem.getIndex(), applicationResultBuildItem.getResult());
+
+                    // add endpoints
+                    for (MethodInfo classEndpoint : classEndpoints) {
+                        if (!hasSecurityAnnotation(classEndpoint)) {
+                            methods.add(classEndpoint);
+                        }
+                    }
                 }
             }
 
-            additionalSecuredClasses.produce(new AdditionalSecuredClassesBuildItem(classes));
-        } else if (securityConfig.defaultRolesAllowed().isPresent() && resteasyDeployment.isPresent()) {
-            List<ClassInfo> classes = new ArrayList<>();
-            Set<DotName> resourceClasses = resteasyDeployment.get().getResult().getScannedResourcePaths().keySet();
-            for (DotName className : resourceClasses) {
-                ClassInfo classInfo = index.getIndex().getClassByName(className);
-                if (!SecurityTransformerUtils.hasSecurityAnnotation(classInfo)) {
-                    classes.add(classInfo);
+            if (!methods.isEmpty()) {
+                if (securityConfig.denyJaxRs()) {
+                    additionalSecuredClasses.produce(new AdditionalSecuredMethodsBuildItem(methods));
+                } else {
+                    additionalSecuredClasses
+                            .produce(new AdditionalSecuredMethodsBuildItem(methods, securityConfig.defaultRolesAllowed()));
                 }
             }
-            additionalSecuredClasses
-                    .produce(new AdditionalSecuredClassesBuildItem(classes, securityConfig.defaultRolesAllowed()));
         }
     }
 

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/DefaultRolesAllowedJaxRsTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/DefaultRolesAllowedJaxRsTest.java
@@ -3,6 +3,7 @@ package io.quarkus.resteasy.reactive.server.test.security;
 import static io.restassured.RestAssured.given;
 import static io.restassured.RestAssured.when;
 
+import org.hamcrest.Matchers;
 import org.jboss.shrinkwrap.api.asset.StringAsset;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -19,7 +20,7 @@ public class DefaultRolesAllowedJaxRsTest {
                     .addClasses(PermitAllResource.class, UnsecuredResource.class,
                             TestIdentityProvider.class,
                             TestIdentityController.class,
-                            UnsecuredSubResource.class)
+                            UnsecuredSubResource.class, HelloResource.class)
                     .addAsResource(new StringAsset("quarkus.security.jaxrs.default-roles-allowed=admin\n"),
                             "application.properties"));
 
@@ -75,6 +76,15 @@ public class DefaultRolesAllowedJaxRsTest {
     public void shouldAllowPermitAllClass() {
         String path = "/permitAll/sub/subMethod";
         assertStatus(path, 200, 200, 200);
+    }
+
+    @Test
+    public void testServerExceptionMapper() {
+        given()
+                .get("/hello")
+                .then()
+                .statusCode(200)
+                .body(Matchers.equalTo("unauthorizedExceptionMapper"));
     }
 
     private void assertStatus(String path, int adminStatus, int userStatus, int anonStatus) {

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/DenyAllJaxRsTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/DenyAllJaxRsTest.java
@@ -3,6 +3,7 @@ package io.quarkus.resteasy.reactive.server.test.security;
 import static io.restassured.RestAssured.given;
 import static io.restassured.RestAssured.when;
 
+import org.hamcrest.Matchers;
 import org.jboss.shrinkwrap.api.asset.StringAsset;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -22,7 +23,7 @@ public class DenyAllJaxRsTest {
                     .addClasses(PermitAllResource.class, UnsecuredResource.class,
                             TestIdentityProvider.class,
                             TestIdentityController.class,
-                            UnsecuredSubResource.class)
+                            UnsecuredSubResource.class, HelloResource.class)
                     .addAsResource(new StringAsset("quarkus.security.jaxrs.deny-unannotated-endpoints = true\n"),
                             "application.properties"));
 
@@ -78,6 +79,15 @@ public class DenyAllJaxRsTest {
     public void shouldAllowPermitAllClass() {
         String path = "/permitAll/sub/subMethod";
         assertStatus(path, 200, 200);
+    }
+
+    @Test
+    public void testServerExceptionMapper() {
+        given()
+                .get("/hello")
+                .then()
+                .statusCode(200)
+                .body(Matchers.equalTo("unauthorizedExceptionMapper"));
     }
 
     private void assertStatus(String path, int status, int anonStatus) {

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/HelloResource.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/HelloResource.java
@@ -1,0 +1,22 @@
+package io.quarkus.resteasy.reactive.server.test.security;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+
+import org.jboss.resteasy.reactive.server.ServerExceptionMapper;
+
+import io.quarkus.security.UnauthorizedException;
+
+@Path("/hello")
+public class HelloResource {
+    @GET
+    public String hello() {
+        return "hello";
+    }
+
+    @ServerExceptionMapper
+    public Response unauthorizedExceptionMapper(UnauthorizedException unauthorizedException) {
+        return Response.ok("unauthorizedExceptionMapper").build();
+    }
+}

--- a/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/AdditionalDenyingUnannotatedTransformer.java
+++ b/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/AdditionalDenyingUnannotatedTransformer.java
@@ -1,5 +1,6 @@
 package io.quarkus.security.deployment;
 
+import static io.quarkus.security.deployment.SecurityProcessor.createMethodDescription;
 import static io.quarkus.security.deployment.SecurityTransformerUtils.DENY_ALL;
 
 import java.util.Collection;
@@ -9,24 +10,25 @@ import java.util.Set;
 import org.jboss.jandex.AnnotationTarget;
 
 import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.quarkus.security.spi.runtime.MethodDescription;
 
 public class AdditionalDenyingUnannotatedTransformer implements AnnotationsTransformer {
 
-    private final Set<String> classNames;
+    private final Set<MethodDescription> methods;
 
-    public AdditionalDenyingUnannotatedTransformer(Collection<String> classNames) {
-        this.classNames = new HashSet<>(classNames);
+    public AdditionalDenyingUnannotatedTransformer(Collection<MethodDescription> methods) {
+        this.methods = new HashSet<>(methods);
     }
 
     @Override
     public boolean appliesTo(AnnotationTarget.Kind kind) {
-        return kind == org.jboss.jandex.AnnotationTarget.Kind.CLASS;
+        return kind == AnnotationTarget.Kind.METHOD;
     }
 
     @Override
     public void transform(TransformationContext context) {
-        String className = context.getTarget().asClass().name().toString();
-        if (classNames.contains(className)) {
+        MethodDescription methodDescription = createMethodDescription(context.getTarget().asMethod());
+        if (methods.contains(methodDescription)) {
             context.transform().add(DENY_ALL).done();
         }
     }

--- a/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/AdditionalRolesAllowedTransformer.java
+++ b/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/AdditionalRolesAllowedTransformer.java
@@ -1,5 +1,6 @@
 package io.quarkus.security.deployment;
 
+import static io.quarkus.security.deployment.SecurityProcessor.createMethodDescription;
 import static io.quarkus.security.deployment.SecurityTransformerUtils.ROLES_ALLOWED;
 
 import java.util.Collection;
@@ -11,27 +12,28 @@ import org.jboss.jandex.AnnotationTarget;
 import org.jboss.jandex.AnnotationValue;
 
 import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.quarkus.security.spi.runtime.MethodDescription;
 
 public class AdditionalRolesAllowedTransformer implements AnnotationsTransformer {
 
-    private final Set<String> classNames;
+    private final Set<MethodDescription> methods;
     private final AnnotationValue[] rolesAllowed;
 
-    public AdditionalRolesAllowedTransformer(Collection<String> classNames, List<String> rolesAllowed) {
-        this.classNames = new HashSet<>(classNames);
+    public AdditionalRolesAllowedTransformer(Collection<MethodDescription> methods, List<String> rolesAllowed) {
+        this.methods = new HashSet<>(methods);
         this.rolesAllowed = rolesAllowed.stream().map(s -> AnnotationValue.createStringValue("", s))
                 .toArray(AnnotationValue[]::new);
     }
 
     @Override
     public boolean appliesTo(AnnotationTarget.Kind kind) {
-        return kind == AnnotationTarget.Kind.CLASS;
+        return kind == AnnotationTarget.Kind.METHOD;
     }
 
     @Override
     public void transform(TransformationContext context) {
-        String className = context.getTarget().asClass().name().toString();
-        if (classNames.contains(className)) {
+        MethodDescription method = createMethodDescription(context.getTarget().asMethod());
+        if (methods.contains(method)) {
             context.transform().add(ROLES_ALLOWED, AnnotationValue.createArrayValue("value", rolesAllowed)).done();
         }
     }

--- a/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java
+++ b/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java
@@ -1,6 +1,8 @@
 package io.quarkus.security.deployment;
 
 import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
+import static io.quarkus.security.deployment.DotNames.DENY_ALL;
+import static io.quarkus.security.deployment.DotNames.ROLES_ALLOWED;
 import static io.quarkus.security.runtime.SecurityProviderUtils.findProviderIndex;
 
 import java.io.IOException;
@@ -77,8 +79,10 @@ import io.quarkus.security.runtime.interceptor.SecurityCheckStorageBuilder;
 import io.quarkus.security.runtime.interceptor.SecurityConstrainer;
 import io.quarkus.security.runtime.interceptor.SecurityHandler;
 import io.quarkus.security.spi.AdditionalSecuredClassesBuildItem;
+import io.quarkus.security.spi.AdditionalSecuredMethodsBuildItem;
 import io.quarkus.security.spi.runtime.AuthorizationController;
 import io.quarkus.security.spi.runtime.DevModeDisabledAuthorizationController;
+import io.quarkus.security.spi.runtime.MethodDescription;
 import io.quarkus.security.spi.runtime.SecurityCheck;
 import io.quarkus.security.spi.runtime.SecurityCheckStorage;
 
@@ -415,6 +419,26 @@ public class SecurityProcessor {
         beans.produce(new AdditionalBeanBuildItem(SecurityHandler.class, SecurityConstrainer.class));
     }
 
+    /**
+     * Transform deprecated {@link AdditionalSecuredClassesBuildItem} to {@link AdditionalSecuredMethodsBuildItem}.
+     */
+    @BuildStep
+    void transformAdditionalSecuredClassesToMethods(List<AdditionalSecuredClassesBuildItem> additionalSecuredClassesBuildItems,
+            BuildProducer<AdditionalSecuredMethodsBuildItem> additionalSecuredMethodsBuildItemBuildProducer) {
+        for (AdditionalSecuredClassesBuildItem additionalSecuredClassesBuildItem : additionalSecuredClassesBuildItems) {
+            final Collection<MethodInfo> securedMethods = new ArrayList<>();
+            for (ClassInfo additionalSecuredClass : additionalSecuredClassesBuildItem.additionalSecuredClasses) {
+                for (MethodInfo method : additionalSecuredClass.methods()) {
+                    if (isPublicNonStaticNonConstructor(method)) {
+                        securedMethods.add(method);
+                    }
+                }
+            }
+            additionalSecuredMethodsBuildItemBuildProducer.produce(
+                    new AdditionalSecuredMethodsBuildItem(securedMethods, additionalSecuredClassesBuildItem.rolesAllowed));
+        }
+    }
+
     /*
      * The annotation store is not meant to be generally supported for security annotation.
      * It is only used here in order to be able to register the DenyAllInterceptor for
@@ -422,21 +446,21 @@ public class SecurityProcessor {
      */
     @BuildStep
     void transformSecurityAnnotations(BuildProducer<AnnotationsTransformerBuildItem> transformers,
-            List<AdditionalSecuredClassesBuildItem> additionalSecuredClasses,
+            List<AdditionalSecuredMethodsBuildItem> additionalSecuredMethods,
             SecurityBuildTimeConfig config) {
         if (config.denyUnannotated) {
             transformers.produce(new AnnotationsTransformerBuildItem(new DenyingUnannotatedTransformer()));
         }
-        if (!additionalSecuredClasses.isEmpty()) {
-            for (AdditionalSecuredClassesBuildItem securedClasses : additionalSecuredClasses) {
-                Set<String> additionalSecured = new HashSet<>();
-                for (ClassInfo additionalSecuredClass : securedClasses.additionalSecuredClasses) {
-                    additionalSecured.add(additionalSecuredClass.name().toString());
+        if (!additionalSecuredMethods.isEmpty()) {
+            for (AdditionalSecuredMethodsBuildItem securedMethods : additionalSecuredMethods) {
+                Collection<MethodDescription> additionalSecured = new HashSet<>();
+                for (MethodInfo additionalSecuredMethod : securedMethods.additionalSecuredMethods) {
+                    additionalSecured.add(createMethodDescription(additionalSecuredMethod));
                 }
-                if (securedClasses.rolesAllowed.isPresent()) {
+                if (securedMethods.rolesAllowed.isPresent()) {
                     transformers.produce(
                             new AnnotationsTransformerBuildItem(new AdditionalRolesAllowedTransformer(additionalSecured,
-                                    securedClasses.rolesAllowed.get())));
+                                    securedMethods.rolesAllowed.get())));
                 } else {
                     transformers.produce(
                             new AnnotationsTransformerBuildItem(
@@ -451,23 +475,22 @@ public class SecurityProcessor {
     void gatherSecurityChecks(BuildProducer<SyntheticBeanBuildItem> syntheticBeans,
             BeanArchiveIndexBuildItem beanArchiveBuildItem,
             BuildProducer<ApplicationClassPredicateBuildItem> classPredicate,
-            List<AdditionalSecuredClassesBuildItem> additionalSecuredClasses,
+            List<AdditionalSecuredMethodsBuildItem> additionalSecuredMethods,
             SecurityCheckRecorder recorder,
             List<AdditionalSecurityCheckBuildItem> additionalSecurityChecks, SecurityBuildTimeConfig config) {
         classPredicate.produce(new ApplicationClassPredicateBuildItem(new SecurityCheckStorageAppPredicate()));
 
-        final Map<DotName, AdditionalSecured> additionalSecured = new HashMap<>();
-        for (AdditionalSecuredClassesBuildItem securedClasses : additionalSecuredClasses) {
-            securedClasses.additionalSecuredClasses.forEach(c -> {
-                if (!additionalSecured.containsKey(c.name())) {
-                    additionalSecured.put(c.name(), new AdditionalSecured(c, securedClasses.rolesAllowed));
-                }
-            });
+        final Map<MethodDescription, AdditionalSecured> additionalSecured = new HashMap<>();
+        for (AdditionalSecuredMethodsBuildItem securedMethods : additionalSecuredMethods) {
+            for (MethodInfo m : securedMethods.additionalSecuredMethods) {
+                additionalSecured.putIfAbsent(createMethodDescription(m),
+                        new AdditionalSecured(m, securedMethods.rolesAllowed));
+            }
         }
 
         IndexView index = beanArchiveBuildItem.getIndex();
         Map<MethodInfo, SecurityCheck> securityChecks = gatherSecurityAnnotations(
-                index, additionalSecured, config.denyUnannotated, recorder);
+                index, additionalSecured.values(), config.denyUnannotated, recorder);
         for (AdditionalSecurityCheckBuildItem additionalSecurityCheck : additionalSecurityChecks) {
             securityChecks.put(additionalSecurityCheck.getMethodInfo(),
                     additionalSecurityCheck.getSecurityCheck());
@@ -499,41 +522,48 @@ public class SecurityProcessor {
 
     private Map<MethodInfo, SecurityCheck> gatherSecurityAnnotations(
             IndexView index,
-            Map<DotName, AdditionalSecured> additionalSecuredClasses, boolean denyUnannotated, SecurityCheckRecorder recorder) {
+            Collection<AdditionalSecured> additionalSecuredMethods, boolean denyUnannotated, SecurityCheckRecorder recorder) {
 
         Map<MethodInfo, AnnotationInstance> methodToInstanceCollector = new HashMap<>();
         Map<ClassInfo, AnnotationInstance> classAnnotations = new HashMap<>();
         Map<MethodInfo, SecurityCheck> result = new HashMap<>(gatherSecurityAnnotations(
-                index, DotNames.ROLES_ALLOWED, methodToInstanceCollector, classAnnotations,
+                index, ROLES_ALLOWED, methodToInstanceCollector, classAnnotations,
                 (instance -> recorder.rolesAllowed(instance.value().asStringArray()))));
         result.putAll(gatherSecurityAnnotations(index, DotNames.PERMIT_ALL, methodToInstanceCollector, classAnnotations,
                 (instance -> recorder.permitAll())));
         result.putAll(gatherSecurityAnnotations(index, DotNames.AUTHENTICATED, methodToInstanceCollector, classAnnotations,
                 (instance -> recorder.authenticated())));
 
-        result.putAll(gatherSecurityAnnotations(index, DotNames.DENY_ALL, methodToInstanceCollector, classAnnotations,
+        result.putAll(gatherSecurityAnnotations(index, DENY_ALL, methodToInstanceCollector, classAnnotations,
                 (instance -> recorder.denyAll())));
 
         /*
-         * Handle additional secured classes by adding the denyAll check to all public non-static methods
-         * that don't have security annotations
+         * Handle additional secured methods by adding the denyAll/rolesAllowed check to all public non-static methods
+         * that don't have same security annotations
          */
-        for (Map.Entry<DotName, AdditionalSecured> additionalSecureClassInfo : additionalSecuredClasses.entrySet()) {
-            for (MethodInfo methodInfo : additionalSecureClassInfo.getValue().classInfo.methods()) {
-                if (!isPublicNonStaticNonConstructor(methodInfo)) {
-                    continue;
+        for (AdditionalSecured additionalSecuredMethod : additionalSecuredMethods) {
+            if (!isPublicNonStaticNonConstructor(additionalSecuredMethod.methodInfo)) {
+                continue;
+            }
+            AnnotationInstance alreadyExistingInstance = methodToInstanceCollector.get(additionalSecuredMethod.methodInfo);
+            if (additionalSecuredMethod.rolesAllowed.isPresent()) {
+                if (alreadyExistingInstance == null) {
+                    result.put(additionalSecuredMethod.methodInfo, recorder
+                            .rolesAllowed(additionalSecuredMethod.rolesAllowed.get().toArray(String[]::new)));
+                } else if (alreadyHasAnnotation(alreadyExistingInstance, ROLES_ALLOWED)) {
+                    // we should not try to add second @RolesAllowed
+                    throw new IllegalStateException("Method " + additionalSecuredMethod.methodInfo.declaringClass() + "#"
+                            + additionalSecuredMethod.methodInfo.name() + " should not have been added as an additional "
+                            + "secured method as it's already annotated with @RolesAllowed.");
                 }
-                AnnotationInstance alreadyExistingInstance = methodToInstanceCollector.get(methodInfo);
-                if ((alreadyExistingInstance == null)) {
-                    if (additionalSecureClassInfo.getValue().rolesAllowed.isPresent()) {
-                        result.put(methodInfo, recorder
-                                .rolesAllowed(additionalSecureClassInfo.getValue().rolesAllowed.get().toArray(String[]::new)));
-                    } else {
-                        result.put(methodInfo, recorder.denyAll());
-                    }
-                } else if (alreadyExistingInstance.target().kind() == AnnotationTarget.Kind.CLASS) {
-                    throw new IllegalStateException("Class " + methodInfo.declaringClass()
-                            + " should not have been added as an additional secured class");
+            } else {
+                if (alreadyExistingInstance == null) {
+                    result.put(additionalSecuredMethod.methodInfo, recorder.denyAll());
+                } else if (alreadyHasAnnotation(alreadyExistingInstance, DENY_ALL)) {
+                    // we should not try to add second @DenyAll
+                    throw new IllegalStateException("Method " + additionalSecuredMethod.methodInfo.declaringClass() + "#"
+                            + additionalSecuredMethod.methodInfo.name() + " should not have been added as an additional "
+                            + "secured method as it's already annotated with @DenyAll.");
                 }
             }
         }
@@ -561,6 +591,11 @@ public class SecurityProcessor {
         }
 
         return result;
+    }
+
+    private boolean alreadyHasAnnotation(AnnotationInstance alreadyExistingInstance, DotName annotationName) {
+        return alreadyExistingInstance.target().kind() == AnnotationTarget.Kind.METHOD
+                && alreadyExistingInstance.name().equals(annotationName);
     }
 
     private boolean isPublicNonStaticNonConstructor(MethodInfo methodInfo) {
@@ -638,13 +673,22 @@ public class SecurityProcessor {
         return AdditionalBeanBuildItem.builder().addBeanClass(controllerClass).build();
     }
 
+    static MethodDescription createMethodDescription(MethodInfo additionalSecuredMethod) {
+        String[] paramTypes = new String[additionalSecuredMethod.parametersCount()];
+        for (int i = 0; i < additionalSecuredMethod.parametersCount(); i++) {
+            paramTypes[i] = additionalSecuredMethod.parameterTypes().get(i).name().toString();
+        }
+        return new MethodDescription(additionalSecuredMethod.declaringClass().name().toString(), additionalSecuredMethod.name(),
+                paramTypes);
+    }
+
     static class AdditionalSecured {
 
-        final ClassInfo classInfo;
+        final MethodInfo methodInfo;
         final Optional<List<String>> rolesAllowed;
 
-        AdditionalSecured(ClassInfo classInfo, Optional<List<String>> rolesAllowed) {
-            this.classInfo = classInfo;
+        AdditionalSecured(MethodInfo methodInfo, Optional<List<String>> rolesAllowed) {
+            this.methodInfo = methodInfo;
             this.rolesAllowed = rolesAllowed;
         }
     }

--- a/extensions/security/spi/src/main/java/io/quarkus/security/spi/AdditionalSecuredClassesBuildItem.java
+++ b/extensions/security/spi/src/main/java/io/quarkus/security/spi/AdditionalSecuredClassesBuildItem.java
@@ -11,6 +11,8 @@ import io.quarkus.builder.item.MultiBuildItem;
 
 /**
  * Contains classes that need to have @DenyAll on all methods that don't have security annotations
+ *
+ * @deprecated use {@link AdditionalSecuredMethodsBuildItem}
  */
 public final class AdditionalSecuredClassesBuildItem extends MultiBuildItem {
 

--- a/extensions/security/spi/src/main/java/io/quarkus/security/spi/AdditionalSecuredMethodsBuildItem.java
+++ b/extensions/security/spi/src/main/java/io/quarkus/security/spi/AdditionalSecuredMethodsBuildItem.java
@@ -1,0 +1,31 @@
+package io.quarkus.security.spi;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import org.jboss.jandex.MethodInfo;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * Contains methods that need to have {@link javax.annotation.security.DenyAll} or
+ * {@link javax.annotation.security.RolesAllowed}.
+ */
+public final class AdditionalSecuredMethodsBuildItem extends MultiBuildItem {
+
+    public final Collection<MethodInfo> additionalSecuredMethods;
+    public final Optional<List<String>> rolesAllowed;
+
+    public AdditionalSecuredMethodsBuildItem(Collection<MethodInfo> additionalSecuredMethods) {
+        this.additionalSecuredMethods = Collections.unmodifiableCollection(additionalSecuredMethods);
+        rolesAllowed = Optional.empty();
+    }
+
+    public AdditionalSecuredMethodsBuildItem(Collection<MethodInfo> additionalSecuredMethods,
+            Optional<List<String>> rolesAllowed) {
+        this.additionalSecuredMethods = Collections.unmodifiableCollection(additionalSecuredMethods);
+        this.rolesAllowed = rolesAllowed;
+    }
+}

--- a/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
@@ -356,6 +356,86 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
     protected List<ResourceMethod> createEndpoints(ClassInfo currentClassInfo,
             ClassInfo actualEndpointInfo, Set<String> seenMethods, Set<String> existingClassNameBindings,
             Set<String> pathParameters, String resourceClassPath, boolean considerApplication) {
+
+        List<ResourceMethod> ret = new ArrayList<>();
+
+        List<FoundEndpoint> endpoints = collectEndpoints(currentClassInfo, actualEndpointInfo, seenMethods,
+                existingClassNameBindings, considerApplication, httpAnnotationToMethod, index, applicationScanningResult,
+                getAnnotationStore());
+        Map<String, BasicResourceClassInfo> basicResourceClassInfoMap = new HashMap<>();
+        for (FoundEndpoint endpoint : endpoints) {
+
+            // assemble method path
+            String methodPath = readStringValue(getAnnotationStore().getAnnotation(endpoint.methodInfo, PATH));
+            if (endpoint.httpMethod != null) {
+                // HTTP annotation method
+                validateHttpAnnotations(endpoint.methodInfo);
+                if (methodPath != null) {
+                    if (!methodPath.startsWith("/")) {
+                        methodPath = "/" + methodPath;
+                    }
+                    if (methodPath.endsWith("/")) {
+                        methodPath = handleTrailingSlash(methodPath);
+                    }
+                } else {
+                    methodPath = "";
+                }
+            } else {
+                // resource locator method
+                if (methodPath != null) {
+                    if (!methodPath.startsWith("/")) {
+                        methodPath = "/" + methodPath;
+                    }
+                    if (methodPath.endsWith("/")) {
+                        methodPath = methodPath.substring(0, methodPath.length() - 1);
+                    }
+                }
+            }
+
+            // create BasicResourceClassInfo and remember it
+            BasicResourceClassInfo basicResourceClassInfo = basicResourceClassInfoMap.get(endpoint.classInfo.name().toString());
+            if (basicResourceClassInfo == null) {
+                String[] classProduces = extractProducesConsumesValues(
+                        getAnnotationStore().getAnnotation(endpoint.classInfo, PRODUCES));
+                String[] classConsumes = extractProducesConsumesValues(
+                        getAnnotationStore().getAnnotation(endpoint.classInfo, CONSUMES));
+                basicResourceClassInfo = new BasicResourceClassInfo(resourceClassPath, classProduces,
+                        classConsumes, pathParameters, getStreamAnnotationValue(endpoint.classInfo));
+                basicResourceClassInfoMap.put(endpoint.classInfo.name().toString(), basicResourceClassInfo);
+            }
+
+            ResourceMethod method = createResourceMethod(endpoint.classInfo, actualEndpointInfo, basicResourceClassInfo,
+                    endpoint.classNameBindings, endpoint.httpMethod, endpoint.methodInfo, methodPath);
+            ret.add(method);
+        }
+
+        return ret;
+    }
+
+    /**
+     * Return endpoints defined directly on classInfo.
+     *
+     * @param classInfo resource class
+     * @return classInfo endpoint method info
+     */
+    public static Collection<MethodInfo> collectClassEndpoints(ClassInfo classInfo,
+            Map<DotName, String> httpAnnotationToMethod, IndexView index, ApplicationScanningResult applicationScanningResult) {
+        Collection<FoundEndpoint> endpoints = collectEndpoints(classInfo, classInfo, new HashSet<>(), new HashSet<>(), true,
+                httpAnnotationToMethod, index, applicationScanningResult, new AnnotationStore(null));
+        Collection<MethodInfo> ret = new HashSet<>();
+        for (FoundEndpoint endpoint : endpoints) {
+            if (endpoint.classInfo.equals(classInfo)) {
+                ret.add(endpoint.methodInfo);
+            }
+        }
+        return ret;
+    }
+
+    private static List<FoundEndpoint> collectEndpoints(ClassInfo currentClassInfo, ClassInfo actualEndpointInfo,
+            Set<String> seenMethods, Set<String> existingClassNameBindings, boolean considerApplication,
+            Map<DotName, String> httpAnnotationToMethod, IndexView index, ApplicationScanningResult applicationScanningResult,
+            AnnotationStore annotationStore) {
+
         if (considerApplication && applicationScanningResult != null
                 && !applicationScanningResult.keepClass(actualEndpointInfo.name().toString())) {
             return Collections.emptyList();
@@ -366,15 +446,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
             return Collections.emptyList();
         }
 
-        List<ResourceMethod> ret = new ArrayList<>();
-        String[] classProduces = extractProducesConsumesValues(getAnnotationStore().getAnnotation(currentClassInfo, PRODUCES));
-        String[] classConsumes = extractProducesConsumesValues(getAnnotationStore().getAnnotation(currentClassInfo, CONSUMES));
-
-        String classStreamElementType = getStreamAnnotationValue(currentClassInfo);
-
-        BasicResourceClassInfo basicResourceClassInfo = new BasicResourceClassInfo(resourceClassPath, classProduces,
-                classConsumes, pathParameters, classStreamElementType);
-
+        List<FoundEndpoint> ret = new ArrayList<>();
         Set<String> classNameBindings = NameBindingUtil.nameBindingNames(index, currentClassInfo);
         if (classNameBindings.isEmpty()) {
             classNameBindings = existingClassNameBindings;
@@ -383,39 +455,25 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
         for (DotName httpMethod : httpAnnotationToMethod.keySet()) {
             List<MethodInfo> methods = currentClassInfo.methods();
             for (MethodInfo info : methods) {
-                AnnotationInstance annotation = getAnnotationStore().getAnnotation(info, httpMethod);
+                AnnotationInstance annotation = annotationStore.getAnnotation(info, httpMethod);
                 if (annotation != null) {
                     if (!hasProperModifiers(info)) {
                         continue;
                     }
-                    validateHttpAnnotations(info);
                     String descriptor = methodDescriptor(info);
                     if (seenMethods.contains(descriptor)) {
                         continue;
                     }
                     seenMethods.add(descriptor);
-                    String methodPath = readStringValue(getAnnotationStore().getAnnotation(info, PATH));
-                    if (methodPath != null) {
-                        if (!methodPath.startsWith("/")) {
-                            methodPath = "/" + methodPath;
-                        }
-                        if (methodPath.endsWith("/")) {
-                            methodPath = handleTrailingSlash(methodPath);
-                        }
-                    } else {
-                        methodPath = "";
-                    }
-                    ResourceMethod method = createResourceMethod(currentClassInfo, actualEndpointInfo,
-                            basicResourceClassInfo, classNameBindings, httpMethod, info, methodPath);
-
-                    ret.add(method);
+                    ret.add(new FoundEndpoint(currentClassInfo, classNameBindings, info, httpMethod));
                 }
             }
         }
+
         //now resource locator methods
         List<MethodInfo> methods = currentClassInfo.methods();
         for (MethodInfo info : methods) {
-            AnnotationInstance annotation = getAnnotationStore().getAnnotation(info, PATH);
+            AnnotationInstance annotation = annotationStore.getAnnotation(info, PATH);
             if (annotation != null) {
                 if (!hasProperModifiers(info)) {
                     continue;
@@ -425,18 +483,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
                     continue;
                 }
                 seenMethods.add(descriptor);
-                String methodPath = readStringValue(annotation);
-                if (methodPath != null) {
-                    if (!methodPath.startsWith("/")) {
-                        methodPath = "/" + methodPath;
-                    }
-                    if (methodPath.endsWith("/")) {
-                        methodPath = methodPath.substring(0, methodPath.length() - 1);
-                    }
-                }
-                ResourceMethod method = createResourceMethod(currentClassInfo, actualEndpointInfo,
-                        basicResourceClassInfo, classNameBindings, null, info, methodPath);
-                ret.add(method);
+                ret.add(new FoundEndpoint(currentClassInfo, classNameBindings, info, null));
             }
         }
 
@@ -444,16 +491,16 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
         if (superClassName != null && !superClassName.equals(OBJECT)) {
             ClassInfo superClass = index.getClassByName(superClassName);
             if (superClass != null) {
-                ret.addAll(createEndpoints(superClass, actualEndpointInfo, seenMethods, classNameBindings,
-                        pathParameters, resourceClassPath, considerApplication));
+                ret.addAll(collectEndpoints(superClass, actualEndpointInfo, seenMethods, classNameBindings, considerApplication,
+                        httpAnnotationToMethod, index, applicationScanningResult, annotationStore));
             }
         }
         List<DotName> interfaces = currentClassInfo.interfaceNames();
         for (DotName i : interfaces) {
             ClassInfo superClass = index.getClassByName(i);
             if (superClass != null) {
-                ret.addAll(createEndpoints(superClass, actualEndpointInfo, seenMethods, classNameBindings,
-                        pathParameters, resourceClassPath, considerApplication));
+                ret.addAll(collectEndpoints(superClass, actualEndpointInfo, seenMethods, classNameBindings, considerApplication,
+                        httpAnnotationToMethod, index, applicationScanningResult, annotationStore));
             }
         }
         return ret;
@@ -481,7 +528,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
         }
     }
 
-    private boolean hasProperModifiers(MethodInfo info) {
+    private static boolean hasProperModifiers(MethodInfo info) {
         if (isSynthetic(info.flags())) {
             log.debug("Method '" + info.name() + " of Resource class '" + info.declaringClass().name()
                     + "' is a synthetic method and will therefore be ignored");
@@ -500,7 +547,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
         return true;
     }
 
-    private boolean isSynthetic(int mod) {
+    private static boolean isSynthetic(int mod) {
         return (mod & 0x1000) != 0; //0x1000 == SYNTHETIC
     }
 
@@ -1707,5 +1754,19 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
      */
     public interface MultipartParameterIndexerExtension {
         void handleMultipartParameter(ClassInfo multipartClassInfo, IndexView index);
+    }
+
+    private static final class FoundEndpoint {
+        private final ClassInfo classInfo;
+        private final Set<String> classNameBindings;
+        private final MethodInfo methodInfo;
+        private final DotName httpMethod;
+
+        private FoundEndpoint(ClassInfo classInfo, Set<String> classNameBindings, MethodInfo methodInfo, DotName httpMethod) {
+            this.classInfo = classInfo;
+            this.classNameBindings = classNameBindings;
+            this.methodInfo = methodInfo;
+            this.httpMethod = httpMethod;
+        }
     }
 }

